packages/sitecore-jss/src/feaas/utils.ts

```diff
@@ -0,0 +1,61 @@
+import {
+  ComponentRendering,
+  HtmlElementRendering,
+  LayoutServiceData,
+  RouteData,
+  reduceLayout,
+} from '../layout';
+
+/**
+ * Walks through rendering tree and returns list of ids of all FEAAS Component Libraries that are used
+ * @param {LayoutServiceData | RouteData<any> | ComponentRendering | HtmlElementRendering | null } object Layout, route data or rendering
+ */
+export function getFEAASLibraryIds(
+  object: LayoutServiceData | RouteData<any> | ComponentRendering | HtmlElementRendering | null
+) {
+  return reduceLayout(
+    object,
+    (initialValue, node) => {
+      let libraryId: string | undefined = undefined;
+
+      if ('fields' in node && node.fields !== undefined) {
+        // Read out FEAASComponent's libraryid
+        if (node.fields.LibraryId) {
+          if ('value' in node.fields.LibraryId && typeof node.fields.LibraryId.value === 'string') {
+            libraryId = node.fields.LibraryId.value;
+          }
+        }
+        // Check CSSStyles override
+        if (node.fields.CSSStyles) {
+          if ('value' in node.fields.CSSStyles && typeof node.fields.CSSStyles.value === 'string') {
+            libraryId = node.fields.CSSStyles.value.match(/-library--([^\s]+)/)?.[1] || libraryId;
```

Comment:
Not sure I understand the fallback to `libraryId` here. Is it expected we could potentially have one set from FEAASComponent's LibraryId field in the code directly above? If so, then wouldn't we want both?

Reply:
It's a good question. Basically, a component relies on a stylesheet of its library. But the user can override the theme of a component, potentially from another library. So in that case we ignore the LibraryId field, and use the one from the class.

Reply:
Makes sense. It's clearer now.